sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -145,249 +804,165 @@ trait GpuHashJoin extends GpuExec {
+    }.map {
+      case (Some(right), Some(left)) => (right.exprId, left)
+      case _ => throw new IllegalStateException("INTERNAL ERROR THIS SHOULD NOT BE REACHABLE")
+    }.toMap
+
+    val rightData = rightOutput.filterNot(att => rightToLeftKeyMap.contains(att.exprId))
+    val remappedRightOutput = rightOutput.map { att =>
+      rightToLeftKeyMap.get(att.exprId)
+          .map(leftAtt => GpuAlias(leftAtt, att.name)(att.exprId))
+          .getOrElse(att)
     }
+    (rightData, remappedRightOutput)
   }
 
   /**
-   * Place the columns in left and the columns in right into a single ColumnarBatch
+   * Spark does joins rather simply. They do it row by row, and as such don't really worry
+   * about how much space is being taken up. We are doing this in batches, and have the option to
+   * deduplicate columns that we know are the same to save even more memory.
+   *
+   * As such we do the join in a few different stages.
+   *
+   * 1. We separate out the join keys from the data that will be gathered. The join keys are used
+   * to produce a gather map, and then can be released. The data needs to stay until it has been
+   * gathered. Depending on the type of join and what is being done the join output is likely to
+   * contain the join keys twice. We don't want to do this because it takes up too much memory
+   * so we remove the keys from the data for one side of the join.
+   *
+   * 2. After this we will do the join. We can produce multiple batches from a single
+   * pair of input batches. The output of this stage is called the intermediate output and is the
+   * data columns each side of the join smashed together.
+   *
+   * 3. In some cases there is a condition that filters out data from the join that should not be
+   * included. In the CPU code the condition will operate on the intermediate output. In some cases
+   * the condition may need to be rewritten to point to the deduplicated key column.
+   *
+   * 4. Finally we need to fix up the data to produce the correct output. This should be a simple
+   * projection that puts the deduplicated keys back to where they need to be.
    */
-  def combine(left: ColumnarBatch, right: ColumnarBatch): ColumnarBatch = {
-    val l = GpuColumnVector.extractColumns(left)
-    val r = GpuColumnVector.extractColumns(right)
-    val c = l ++ r
-    new ColumnarBatch(c.asInstanceOf[Array[ColumnVector]], left.numRows())
+  protected lazy val (leftData, rightData, intermediateOutput, finalProject) = {
+    require(leftKeys.map(_.dataType) == rightKeys.map(_.dataType),
+      "Join keys from two sides should have same types")
+    val (leftData, remappedLeftOutput, rightData, remappedRightOutput) = joinType match {
+      case FullOuter | RightOuter | LeftOuter =>
+        // We cannot dedupe anything here because the we can get nulls in the key columns
+        // at least one side
+        (left.output, left.output, right.output, right.output)
+      case _: InnerLike | LeftSemi | LeftAnti =>
```

Comment:
Unlike inner join, semi and anti joins don't need anything from the right table except the keys.  I see there's special handling of these joins just below, but that's only for the final project.  It looks like we'll still cart around the right table data for these joins.